src/main/java/org/openrewrite/staticanalysis/EmptyBlockVisitor.java

```diff
@@ -223,6 +221,29 @@ private boolean isEmptyBlock(Statement blockNode) {
         return false;
     }
 
+    private boolean isEmptyResources(@Nullable List<J.Try.Resource> resources) {
+        if (resources == null || resources.isEmpty()) {
+            return true;
+        }
+        // Searching for access to instances from outside the scope to detect potential side effects.
+        // If that's the case, we cannot remove this resources block.
+        for (J.Try.Resource resource : resources) {
+            // Any reference to an identifier used here comes from outside the scope.
+            if (resource.getVariableDeclarations() instanceof J.Identifier) {
+                return false;
+            } else if (resource.getVariableDeclarations() instanceof J.VariableDeclarations) {
+                J.VariableDeclarations variableDeclarations = (J.VariableDeclarations) resource.getVariableDeclarations();
+                for (J.VariableDeclarations.NamedVariable variable : variableDeclarations.getVariables()) {
+                    // If the variable is not initialized with a new instance, it means it can come from outside the scope.
+                    if (!(variable.getInitializer() instanceof J.NewClass)) {
+                        return false;
+                    }
+                }
+            }
+        }
+        return true;
```

Comment:
I'm not clear why this is necessary -- if resources is null or empty, that should be good enough. In the case I reported, there will be no other reference to them outside of the resources block because they're coming from an input variable to a lambda. I would rather us make fewer, safer changes, than make more changes and risk breaking.

Reply:
Without this, the existing test https://github.com/openrewrite/rewrite-static-analysis/blob/a58e9c52ad0ec5df6cd059517cd6200e034ebc0f/src/test/java/org/openrewrite/staticanalysis/EmptyBlockTest.java#L109-L134 was failing.

Actually, the code here is limiting the scope of changes, only removing the block on less scenarios than before, but keeping the one that on the tests we agreed that is valid.